Modules/Sources/Yosemite/PointOfSale/Items/GRDBObservableDataSource.swift

```diff
@@ -0,0 +1,232 @@
+    private let itemMapper: PointOfSaleItemMapperProtocol
+    private let pageSize: Int
+
+    private var currentProductPage: Int = 1
+    private var currentVariationPage: Int = 1
+    private var currentParentProduct: POSVariableParentProduct?
+    private var totalProductCount: Int = 0
+    private var totalVariationCount: Int = 0
+
+    // ValueObservation subscriptions
+    private var productObservationCancellable: AnyCancellable?
+    private var variationObservationCancellable: AnyCancellable?
+    private var statisticsObservationCancellable: AnyCancellable?
+
+    // MARK: - Initialization
+
+    public init(siteID: Int64,
+                grdbManager: GRDBManagerProtocol,
+                currencySettings: CurrencySettings,
+                itemMapper: PointOfSaleItemMapperProtocol? = nil,
+                pageSize: Int = 20) {
+        self.siteID = siteID
+        self.grdbManager = grdbManager
+        self.itemMapper = itemMapper ?? PointOfSaleItemMapper(currencySettings: currencySettings)
+        self.pageSize = pageSize
+
+        setupStatisticsObservation()
+    }
+
+    deinit {
+        productObservationCancellable?.cancel()
+        variationObservationCancellable?.cancel()
+        statisticsObservationCancellable?.cancel()
+    }
+
+    // MARK: - POSObservableDataSourceProtocol
+
+    public func loadProducts() {
+        currentProductPage = 1
+        isLoadingProducts = true
+        setupProductObservation()
+    }
+
+    public func loadMoreProducts() {
+        guard hasMoreProducts && !isLoadingProducts else { return }
+
+        isLoadingProducts = true
+        currentProductPage += 1
+        setupProductObservation()
+    }
+
+    public func loadVariations(for parentProduct: POSVariableParentProduct) {
+        guard currentParentProduct?.productID != parentProduct.productID else {
+            return // Same parent - idempotent
+        }
+
+        currentParentProduct = parentProduct
+        currentVariationPage = 1
+        isLoadingVariations = true
+        variationItems = []
+
+        setupVariationObservation(parentProduct: parentProduct)
+    }
+
+    public func loadMoreVariations() {
+        guard let parentProduct = currentParentProduct,
+              hasMoreVariations && !isLoadingVariations else { return }
+
+        isLoadingVariations = true
+        currentVariationPage += 1
+        setupVariationObservation(parentProduct: parentProduct)
+    }
+
+    public func refresh() {
+        // No-op: database observation automatically updates when data changes during incremental sync
+    }
+
+    // MARK: - ValueObservation Setup
+
+    private func setupProductObservation() {
+        let currentPage = currentProductPage
+        let observation = ValueObservation
+            .tracking { [weak self] database -> [POSProduct] in
+                guard let self else { return [] }
+
+                let persistedProducts = try PersistedProduct
```

Comment:
It's a minor detail, but I wonder if this observation and the statistics observation should happen in the same observable to avoid any data inconsistencies. I imagine both updates should happen at the same time, but theoretically, there can be a gap between setting `productItems` and `totalProductCount`, making `hasMoreProducts` value wrong for a brief moment. 🤔